src/Common/src/TypeSystem/Interop/IL/Marshaller.cs

```diff
@@ -56,211 +88,791 @@ protected Marshaller()
             }
 
-            if (type is MetadataType)
+            if (elementType.IsPointer)
             {
-                var metadataType = (MetadataType)type;
+                return MarshallerKind.InvalidUnsafeType;
+            }
 
-                if (metadataType.Module == context.SystemModule)
-                {
-                    var nameSpace = metadataType.Namespace;
-                    var name = metadataType.Name;
+            bool isAnsi = (methodData.GetCharSet() & PInvokeAttributes.CharSetAnsi) == PInvokeAttributes.CharSetAnsi;
+            NativeType nativeType = NativeType.NotSet;
+
+            if (marshalAs != null)
+                nativeType = (NativeType)marshalAs.ArraySubType;
 
-                    if (name == "StringBuilder" && nameSpace == "System.Text")
+            switch (elementType.Category)
+            {
+                case TypeFlags.Char:
+                    switch (nativeType)
                     {
-                        return MarshallerKind.StringBuilder;
+                        case NativeType.I1:
+                        case NativeType.U1:
+                            return MarshallerKind.AnsiChar;
+                        case NativeType.I2:
+                        case NativeType.U2:
+                            return MarshallerKind.UnicodeChar;
+                        default:
+                            if (isAnsi)
+                                return MarshallerKind.AnsiChar;
+                            else
+                                return MarshallerKind.UnicodeChar;
                     }
-                }
+
+                case TypeFlags.Boolean:
+                    switch (nativeType)
+                    {
+                        case NativeType.Boolean:
+                            return MarshallerKind.Bool;
+                        case NativeType.I1:
+                        case NativeType.U1:
+                            return MarshallerKind.CBool;
+                        case NativeType.NotSet:
+                        default:
+                            return MarshallerKind.Bool;
+                    }
+                case TypeFlags.IntPtr:
+                case TypeFlags.UIntPtr:
+                    return MarshallerKind.BlittableValue;
+
+                case TypeFlags.Void:
+                    return MarshallerKind.InvalidUnsafeType;
+
+                case TypeFlags.SByte:
+                case TypeFlags.Int16:
+                case TypeFlags.Int32:
+                case TypeFlags.Int64:
+                case TypeFlags.Byte:
+                case TypeFlags.UInt16:
+                case TypeFlags.UInt32:
+                case TypeFlags.UInt64:
+                case TypeFlags.Single:
+                case TypeFlags.Double:
+                    return MarshallerKind.BlittableValue;
             }
 
-            if (PInvokeMethodData.IsSafeHandle(type))
+            if (elementType.IsString)
```

Comment:
It would be nice to move the check for things that are not value types into the else part of the `IsValueType` check for better perf.